### PR TITLE
Try not to copy when allocating from iterators, take 3

### PR DESCRIPTION
**Pull request by bugadani**
The problem I'm trying to solve here (and in previous PRs) is that, `TypedArena::alloc_from_iter` first creates a vector from the iterator's contents, then copies it into the arena. I'm trying to reuse the allocated vector instead of copying it's contents.

Previous attempts have failed because of either unsoundness (#77945) or bad performance (#78067). This PR tried to address the issue by creating two, parallel arenas as part of a single `TypedArena`: one for "scalar" allocations, which includes slices of known size, and one for "vector" allocation, that store the vector objects created by `alloc_from_iter`. ~While this increases the size of `TypedArena`, I hope this split results in better performance since allocating from an iter means moving the `Vec` object to an already known address, instead of having to compute it through a vector.~ This change also wasn't able to achieve the desired performance, the increased size and complexity also caused slight regressions.

Current state of the PR: I've reverted most of the above changes and only kept special casing the allocation where the type inside `TypedArena` does not need special considerations when dropping.

### CODE REVIEW

**Comment by bugadani** [comment]
Oh, technically, this is a work in progress: I still need to put some time into naming things and clean up, but I don't want to do that before knowing if the change makes sense.

File: compiler/rustc_arena/src/lib.rs [hunk 1]
```diff
@@ -217,8 +237,10 @@ impl<T> TypedArena<T> {
             let mut chunks = self.chunks.borrow_mut();
             let mut new_cap;
             if let Some(last_chunk) = chunks.last_mut() {
-                let used_bytes = self.ptr.get() as usize - last_chunk.start() as usize;
-                last_chunk.entries = used_bytes / mem::size_of::<T>();
+                if mem::needs_drop::<T>() {
```
Comment (oli-obk) [inline compiler/rustc_arena/src/lib.rs:240]: Why is this needed?

File: compiler/rustc_arena/src/lib.rs [hunk 2]
```diff
@@ -282,7 +304,89 @@ impl<T> TypedArena<T> {
...
+    #[inline]
+    pub fn alloc_slice(&self, slice: &[T]) -> &mut [T]
+    where
+        T: Copy,
+    {
+        self.scalar.alloc_slice(slice)
+    }
+
+    #[inline]
+    pub fn alloc_from_iter_general<I: Iterator<Item = T>>(&self, iter: I) -> &mut [T] {
```
Comment (oli-obk) [inline compiler/rustc_arena/src/lib.rs:330]: can this now be invoked with zero sized types? Or should this function be private?

**Comment by oli-obk** [comment]
cc @est31 @Marwes 

@bors try @rust-timer queue

File: compiler/rustc_arena/src/lib.rs (hunk 2 above)
Comment (bugadani) [reply compiler/rustc_arena/src/lib.rs:330]: It should be private

File: compiler/rustc_arena/src/lib.rs (hunk 1 above)
Comment (bugadani) [reply compiler/rustc_arena/src/lib.rs:240]: There are some places where TypedArena is created for non-drop types. It doesn't need to track the entries in that case.

Comment (oli-obk) [reply compiler/rustc_arena/src/lib.rs:240]: That seems like an orthogonal change to this PR that I'd be more comfortable reviewing in a separate PR. Especially since I think there are more places to do this then.

Comment (bugadani) [reply compiler/rustc_arena/src/lib.rs:240]: I only found this one, but I can extract it into a different PR

**Comment by bugadani** [comment]
Here's what I know:
 - duplicating the arena struct like this is definitely not worth it
 - the `!needs_drop` "copy immediately into the arena" can result in small, but measurable improvement, as demonstrated in #77945 - it really sucks that's unsound.
 - the above speedup isn't significant looking at the "big picture"
 - I think `SmallVec` causes a perf hit

I wonder if keeping the `!needs_drop` special case would be worth it. Normally, those should be allocated in `DroplessArena` I think, but the query system uses `TypedArena` because it can handle both, so there is actual usage of `TypedArenas` for no-drop type data.

**Comment by oli-obk** [comment]
So... you're suggesting we dispatch on whether something needs drop even earlier? Like have the check in the query system and then switch to `DroplessArena`?

**Comment by bugadani** [comment]
I think using `DroplessArena` wherever it's possible would indeed be preferrable, but I don't see how that can be done right now.

**Comment by bugadani** [comment]
Well I rebased this patch, but thinking about it more, these changes should have no or negligible effect so I'm closing it.
